File: frame/src/webhook.nim

import json, httpclient, zippy # , queues, threading, times
from ./config import Config

type
  Webhook* = ref object
    config: Config
    # queue: Queue[JsonNode]
    # stopEvent: Event
    # thread: Thread[void]

proc newWebhook*(config: Config): Webhook =
  new(result)
  result.config = config
#   open(result.queue)
#   initEvent(result.stopEvent)
#   createThread(result.thread, result.run)

# method addLog*(self: Webhook, payload: JsonNode) =
#   self.queue.put(payload)

# method run(self: Webhook) =
#   while not self.stopEvent.test:
#     var batch: seq[JsonNode]

#     # Get at least one item, blocking if the queue is empty
#     batch.add(self.queue.get())

#     # Try to fill the batch up to its max size without blocking
#     for i in 0..<99:
#       if not self.queue.isEmpty:
#         batch.add(self.queue.get())

#     self.sendBatch(batch)

# method sendBatch(self: Webhook, batch: seq[JsonNode]) =
method addLog*(self: Webhook, payload: JsonNode) =
  try:
    let client = newHttpClient(timeout = 10000)
    client.headers = newHttpHeaders([
        ("Authorization", "Bearer " & self.config.serverApiKey),
        ("Content-Type", "application/json"),
        ("Content-Encoding", "gzip")
    ])
    let protocol = if self.config.serverPort mod 1000 ==
        443: "https" else: "http"
    let url = protocol & "://" & self.config.serverHost & ":" &
        $self.config.serverPort & "/api/log"

    let body = %*{
        "log": payload
    }

    let response = client.request(url, httpMethod = HttpPost, body = compress($body))
    echo response.status
    # if response.status != Http200:
    #   echo "Error sending logs: HTTP " & $response.status
  except CatchableError as e:
    echo "Error sending logs: " & $e.msg

proc stopWebhook*(self: Webhook) =
#   self.stopEvent.set()
#   joinThread(self.thread)
  echo "Webhook stopped"

File: frameos-python/frame/webhook.py
import gzip
import json

import requests
import logging

from queue import Queue, Empty

from typing import List, Dict, Any
from threading import Thread, Event

from .config import Config

class Webhook:
    def __init__(self, config: Config):
        self.config = config
        self.queue = Queue()
        self.stop_event = Event()
        self.thread = Thread(target=self._run)
        self.thread.start()

    def add_log(self, payload: Dict[str, Any]):
        self.queue.put(payload)

    def _run(self):
        while not self.stop_event.is_set():
            batch = []
            
            # Start by getting at least one item. This will block if the queue is empty.
            item = self.queue.get()
            batch.append(item)

            # Then try to fill the batch up to its max size without blocking
            for _ in range(99):
                try:
                    item = self.queue.get_nowait()
                    batch.append(item)
                except Empty:
                    break

            self._send_batch(batch)

    def _send_batch(self, batch: List[Dict[str, Any]]):
        if not self.config:
            return
        protocol = 'https' if self.config.server_port % 1000 == 443 else 'http'
        url = f"{protocol}://{self.config.server_host}:{self.config.server_port}/api/log"
        headers = {
            "Authorization": f"Bearer {self.config.server_api_key}",
            "Content-Type": "application/json",
            "Content-Encoding": "gzip",
        }
        try:
            payload = json.dumps({"logs": batch}).encode('utf-8')
            response = requests.post(url, headers=headers, data=gzip.compress(payload))
            response.raise_for_status()
        except requests.HTTPError as e:
            logging.error(f"Error sending logs (HTTP {response.status_code}): {e}")
        except Exception as e:
            logging.error(f"Error sending logs: {e}")

    def stop(self):
        self.stop_event.set()
        self.thread.join()
